packages/shared/src/toDisplayString.ts

```diff
@@ -12,11 +13,22 @@ import {
  * @private
  */
 export const toDisplayString = (val: unknown): string => {
-  return val == null
-    ? ''
-    : isArray(val) || (isObject(val) && val.toString === objectToString)
-    ? JSON.stringify(val, replacer, 2)
-    : String(val)
+  if (val == null) return ''
+
+  let hasToString = false
+  if (!isObject(val)) {
+    hasToString = true
+  } else if (isArray(val)) {
+    // skip toString override check for Array
+    hasToString = false
```

Comment:
is this branch necessary?

Reply:
you might be right, I added explicit branches for each possible cases I wanted to cover.
I assumed the terser would handle the optimization. 

I push a simplified version for the code, that I still find clear.